File: taurus-protect-sdk-go/pkg/protect/helper/price_verifier.go
package helper

import (
	"encoding/json"
	"fmt"

	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// canonicalPrice is the projection of a price that a PRICEUPDATER signs: the five fields
// validatord's CurrencyPrice serialises, in its field order. Go's encoding/json emits
// struct fields in declaration order, so this order IS the signed byte sequence — do not
// reorder or add fields.
type canonicalPrice struct {
	Blockchain   string `json:"blockchain"`
	CurrencyFrom string `json:"currencyFrom"`
	CurrencyTo   string `json:"currencyTo"`
	Decimals     string `json:"decimals"`
	Rate         string `json:"rate"`
}

// PriceSignedBytes returns the exact bytes a price signature covers.
func PriceSignedBytes(price *model.Price) ([]byte, error) {
	if price == nil {
		return nil, fmt.Errorf("price cannot be nil")
	}
	return json.Marshal(canonicalPrice{
		Blockchain:   price.Blockchain,
		CurrencyFrom: price.CurrencyFrom,
		CurrencyTo:   price.CurrencyTo,
		Decimals:     price.Decimals,
		Rate:         price.Rate,
	})
}

// VerifyPrice checks a price against the PRICEUPDATER keys in a verified rules container.
//
// The container decides whether prices must be signed at all. It is SuperAdmin-verified,
// so "this tenant has no price signer" is trustworthy, whereas "this price carries no
// signatures" is not — which is why a stripped signatures array on a tenant that DOES
// have a PRICEUPDATER is an error rather than a skip.
//
// Rate and decimals feed amount conversion, so an unverified price is a wrong number a
// caller would otherwise act on.
//
//	verified container ─▶ any PRICEUPDATER key? ──no──▶ PASS (tenant does not sign prices)
//	                              │yes
//	                              ▼
//	                      price.Signatures empty? ──yes──▶ IntegrityError
//	                              │no                      (only the container may excuse
//	                              ▼                         a price from being signed)
//	                      one entry valid over the
//	                      canonical 5-field JSON? ──no──▶ IntegrityError
//	                              │yes
//	                              ▼
//	                             PASS
func VerifyPrice(price *model.Price, rulesContainer *model.DecodedRulesContainer) error {
	if price == nil {
		return &model.IntegrityError{Message: "price cannot be nil"}
	}
	if rulesContainer == nil {
		return &model.IntegrityError{
			Message: "rules container required for price signature verification",
		}
	}

	keys := rulesContainer.PriceUpdaterKeys()
	if len(keys) == 0 {
		// This tenant does not sign prices.
		return nil
	}

	if len(price.Signatures) == 0 {
		return &model.IntegrityError{
			Message: fmt.Sprintf(
				"price %s/%s carries no signatures but the rules container configures a PRICEUPDATER",
				price.CurrencyFrom, price.CurrencyTo),
		}
	}

	data, err := PriceSignedBytes(price)
	if err != nil {
		return &model.IntegrityError{
			Message: fmt.Sprintf("cannot build the signed form of price %s/%s: %v",
				price.CurrencyFrom, price.CurrencyTo, err),
		}
	}

	for _, sig := range price.Signatures {
		if sig.Signature == "" {
			continue
		}
		if IsValidSignature(data, sig.Signature, keys) {
			return nil
		}
	}

	return &model.IntegrityError{
		Message: fmt.Sprintf(
			"no PRICEUPDATER signature verifies for price %s/%s (%d signature(s) offered)",
			price.CurrencyFrom, price.CurrencyTo, len(price.Signatures)),
	}
}

// VerifyPrices verifies each price, returning the first failure.
func VerifyPrices(prices []*model.Price, rulesContainer *model.DecodedRulesContainer) error {
	for _, price := range prices {
		if price == nil {
			continue
		}
		if err := VerifyPrice(price, rulesContainer); err != nil {
			return err
		}
	}
	return nil
}

File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/helper/PriceVerifier.java
package com.taurushq.sdk.protect.client.helper;

import com.google.gson.Gson;
import com.taurushq.sdk.protect.client.model.IntegrityException;
import com.taurushq.sdk.protect.client.model.Price;
import com.taurushq.sdk.protect.client.model.PriceSignature;
import com.taurushq.sdk.protect.client.model.rulescontainer.DecodedRulesContainer;
import com.taurushq.sdk.protect.client.model.rulescontainer.RuleUser;

import java.nio.charset.StandardCharsets;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.List;

/**
 * Price signature verification against the PRICEUPDATER role.
 * <p>
 * Rate and decimals feed amount conversion, so an unverified price is a wrong number a
 * caller acts on.
 */
public final class PriceVerifier {

    private static final Gson GSON = new Gson();
    private static final String PRICE_UPDATER_ROLE = "PRICEUPDATER";

    private PriceVerifier() {
    }

    /**
     * Returns the exact bytes a price signature covers.
     * <p>
     * This is validatord's CurrencyPrice JSON projection: the five fields it serialises,
     * in its field order, with no spaces. The key order IS the signed byte sequence — do
     * not reorder or add fields.
     *
     * @param price the price
     * @return the canonical signed bytes
     */
    public static byte[] priceSignedBytes(final Price price) {
        if (price == null) {
            throw new IllegalArgumentException("price cannot be null");
        }
        String canonical = "{\"blockchain\":" + GSON.toJson(nullToEmpty(price.getBlockchain()))
                + ",\"currencyFrom\":" + GSON.toJson(nullToEmpty(price.getCurrencyFrom()))
                + ",\"currencyTo\":" + GSON.toJson(nullToEmpty(price.getCurrencyTo()))
                + ",\"decimals\":" + GSON.toJson(nullToEmpty(price.getDecimals()))
                + ",\"rate\":" + GSON.toJson(nullToEmpty(price.getRate()))
                + "}";
        return canonical.getBytes(StandardCharsets.UTF_8);
    }

    /**
     * Checks a price against the PRICEUPDATER keys in a verified rules container.
     * <p>
     * The container decides whether prices must be signed at all. It is
     * SuperAdmin-verified, so "this tenant has no price signer" is trustworthy, whereas
     * "this price carries no signatures" is not — which is why a stripped signatures list
     * on a tenant that DOES have a PRICEUPDATER is an error rather than a skip.
     *
     * @param price          the price to verify
     * @param rulesContainer the verified rules container
     * @throws IntegrityException if no PRICEUPDATER signature verifies
     */
    public static void verifyPrice(final Price price, final DecodedRulesContainer rulesContainer) {
        if (price == null) {
            throw new IntegrityException("price cannot be null");
        }
        if (rulesContainer == null) {
            throw new IntegrityException(
                    "rules container required for price signature verification");
        }

        List<PublicKey> keys = priceUpdaterKeys(rulesContainer);
        if (keys.isEmpty()) {
            // This tenant does not sign prices.
            return;
        }

        List<PriceSignature> signatures = price.getSignatures();
        if (signatures == null || signatures.isEmpty()) {
            throw new IntegrityException(String.format(
                    "price %s/%s carries no signatures but the rules container configures a %s",
                    price.getCurrencyFrom(), price.getCurrencyTo(), PRICE_UPDATER_ROLE));
        }

        byte[] data = priceSignedBytes(price);
        for (PriceSignature sig : signatures) {
            if (sig == null || sig.getSignature() == null || sig.getSignature().isEmpty()) {
                continue;
            }
            if (SignatureVerifier.isValidSignature(data, sig.getSignature(), keys)) {
                return;
            }
        }

        throw new IntegrityException(String.format(
                "no %s signature verifies for price %s/%s (%d signature(s) offered)",
                PRICE_UPDATER_ROLE, price.getCurrencyFrom(), price.getCurrencyTo(),
                signatures.size()));
    }

    /**
     * Verifies each price, throwing on the first failure.
     *
     * @param prices         the prices to verify
     * @param rulesContainer the verified rules container
     * @throws IntegrityException if any price fails
     */
    public static void verifyPrices(final List<Price> prices,
                                    final DecodedRulesContainer rulesContainer) {
        if (prices == null) {
            return;
        }
        for (Price price : prices) {
            if (price != null) {
                verifyPrice(price, rulesContainer);
            }
        }
    }

    /**
     * Public keys of every user carrying the PRICEUPDATER role.
     */
    private static List<PublicKey> priceUpdaterKeys(final DecodedRulesContainer rulesContainer) {
        List<PublicKey> keys = new ArrayList<>();
        if (rulesContainer.getUsers() == null) {
            return keys;
        }
        for (RuleUser user : rulesContainer.getUsers()) {
            if (user == null || user.getPublicKey() == null || user.getRoles() == null) {
                continue;
            }
            if (user.getRoles().contains(PRICE_UPDATER_ROLE)) {
                keys.add(user.getPublicKey());
            }
        }
        return keys;
    }

    private static String nullToEmpty(final String value) {
        return value == null ? "" : value;
    }
}
